src/parser/parser_flow.ml

```diff
@@ -2556,7 +2565,18 @@ end = struct
               body;
             }))
         | _ ->
+            raise Try.Rollback
+      in
+
+      fun env ->
+        let start_loc = Peek.loc env in
+        Expect.token env T_FOR;
+        Expect.token env T_LPAREN;
+        match Try.to_parse env (try_forin_or_forof start_loc) with
```

Comment:
Using the `Try` module is a little dangerous for performance, so we need to be pretty judicious with its use. I'm not 100% sure we need it for this case. Can we do something like

1. expect T_FOR
2. expect T_LPAREN
3. parse init declaration
4. Decide which kind of loop to use
  * If the init declaration is initialized, assume for loop
  * else if the next token is `T_OF` assume for-of loop
  * else if the next token is `T_IN` assume for-in loop
  * else assume for loop

If we do need to use `Try`, it's important to make sure we try the most likely path first. I'm guessing that for is more popular than for-of and for-in so we probably would want to try that first. The pattern I'm starting to see around the `Try` module is

1) Try most likely path.
2) If there are errors, try less likely path
3) If there are errors, go back to most likely path, since those errors are more likely the ones we want to show.